### PR TITLE
Add Google reCAPTCHA v3 to online demo

### PR DESCRIPTION
**Pull request by apedroferreira**
- Closes https://github.com/mui/mui-toolpad/issues/1010

Use Google reCAPTCHA v3 to verify users upon creating an app.
This behavior is optional and enabled by the environment variables for the reCAPTCHA.

Google's reCAPTCHA v3 uses a score to try to determine whether a user is a person or a bot - it does not show any puzzle to solve or anything of the sort, which is what captcha v2 used to do. I set the minimum acceptable score as 0.5 (it goes from 0-1), which seems to be recommended as a reasonable default value.

**UI changes:**
Using the CAPTCHA caused a small badge to show up in the lower-right corner of the website. Because I decided to hide it with CSS, it is required to add the following disclaimer in the app creation form when we use the captcha:
https://stackoverflow.com/questions/44543157/how-to-hide-the-google-invisible-recaptcha-badge
Screen Shot 2022-09-26 at 18 35 54

**Create app mutation error changes:**
I've also allowed showing custom errors when creating an app, as the current logic for is it to always show `An app with this name already exists`, no matter what the error is. I've made it so that we have to throw errors with a certain message to be shown.

### CODE REVIEW

File: packages/toolpad-app/src/utils/recaptchaV3.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+): Promise<boolean> => {
+  const recaptchaParams = new URLSearchParams({
+    secret: secretKey,
+    response: token,
+  });
+
+  const recaptchaResponse = await fetch(`${SITE_VERIFY_URL}?${recaptchaParams}`, {
+    method: 'POST',
+  });
+  const recaptchaReponseJson = await recaptchaResponse.json();
```
Comment (bytasv) [inline packages/toolpad-app/src/utils/recaptchaV3.ts:23]: minor suggestion:
```
const { success, score } = await recaptchaResponse.json()

if (!success || score < SCORE_THRESHOLD) {
```

**Review by bytasv — APPROVED** [review]

File: packages/toolpad-app/src/toolpad/Home/index.tsx [hunk 2]
```diff
@@ -124,7 +132,7 @@ function CreateAppDialog({ onClose, ...props }: CreateAppDialogProps) {
               label="Name"
               value={name}
               error={createAppMutation.isError}
-              helperText={createAppMutation.isError ? `An app named "${name}" already exists` : ''}
+              helperText={(createAppMutation.error as Error)?.message || ''}
```
Comment (bharatkashyap) [inline packages/toolpad-app/src/toolpad/Home/index.tsx:135]:
```suggestion
              helperText={(createAppMutation.isError ? (createAppMutation.error as Error)?.message : ''}
```

**Review by bharatkashyap — COMMENTED** [review]
Looks good! I was under the impression that this was only for the demo version, but looks like this is being added across environments. Is that intended?

File: packages/toolpad-app/src/toolpad/Home/index.tsx (hunk 2 above)
Comment (apedroferreira) [reply packages/toolpad-app/src/toolpad/Home/index.tsx:135]: does this have any particular advantage? 
i thought that checking just for the error would be correct, is there any case where that could have issues?

**Comment by apedroferreira** [comment]
> Looks good! I was under the impression that this was only for the demo version, but looks like this is being added across environments. Is that intended?

We can enable or disabled by including the environment variables for the captcha or not - so there will be no captcha unless you add it.
I think that for now the plan is for this feature to be enabled in the demo version of Toolpad only, even though it could technically be a separate feature.

**Comment by oliviertassinari** [comment]
> Google reCAPTCHA v3

I'm curious to see how well this will work.

4 years ago, I was facing an issue with the integration of Stripe. We were processing a lot of $1 payments, for which Stripe detected a lot of bot testing credit cards with. They asked us to implement an anti-boat protection. I started with Google reCAPTCHA v3 but it was blocking actual users and letting bot go thought 😅. We had to move to the v2. On the other hand https://strapi.io/demo uses the v3 (as you can see from the required banner) so 🤷

File: packages/toolpad-app/src/utils/recaptcha.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  await new Promise<void>((resolve) => grecaptcha.ready(resolve));
+  return grecaptcha.execute(siteKey, {
+    action: 'submit',
+  });
+};
+
+const SITE_VERIFY_URL = 'https://www.google.com/recaptcha/api/siteverify';
+const SCORE_THRESHOLD = 0.5;
+
+export const validateRecaptchaToken = async (
```
Comment (Janpot) [inline packages/toolpad-app/src/utils/recaptcha.ts:11]: @apedroferreira This file is mixing serverside-only code with client side code.   You might want to split that

Comment (apedroferreira) [reply packages/toolpad-app/src/utils/recaptcha.ts:11]: added the changes here https://github.com/mui/mui-toolpad/pull/1087/commits/8ec6c3dbff03c1e4201a1be183e12c0db71a87d1

**Pull request merged**